nautobot/extras/models/groups.py

```diff
@@ -566,43 +558,57 @@ def members(self):
+        #       but bear in mind that this just creates the queryset, it doesn't actually evaluate it here.
+        #       Updating the cache would require actually evaluating the queryset, which could be relatively expensive,
+        #       and might actually be unnecessary (e.g. if we are doing `group.members.filter(...)` we don't need
+        #       to evaluate `group.members` itself).
+        #       Is there a way we could have the cache auto-update only when the queryset is evaluated?
+        # return self.update_cached_members(members=members)
+        return members
 
     @property
     def members_cache_key(self):
-        """Return the cache key for this group's members."""
+        """Obsolete cache key for this group's members."""
         return f"nautobot.extras.dynamicgroup.{self.id}.members_cached"
 
     @property
     def members_cached(self):
         """Return the member objects for this group, cached if available."""
 
-        unpickled_query = None
         try:
-            cached_query = cache.get(self.members_cache_key)
-            if cached_query is not None:
-                unpickled_query = pickle.loads(cached_query)  # noqa: S301  # suspicious-pickle-usage -- we know, but we control what's in the DB
-        except pickle.UnpicklingError:
-            logger.warning("Failed to unpickle cached members for %s", self)
-        finally:
-            if unpickled_query is None:
-                unpickled_query = self.members.all()
-                cached_query = pickle.dumps(unpickled_query)  # Explicitly pickle the query to evaluate it.
-                cache.set(
-                    self.members_cache_key, cached_query, get_settings_or_config("DYNAMIC_GROUPS_MEMBER_CACHE_TIMEOUT")
-                )
-
-        return unpickled_query
-
-    def update_cached_members(self):
+            return self._backing_group.members
+        except ObjectDoesNotExist:
+            return self.update_cached_members()
+
+    def update_cached_members(self, members=None):
         """
         Update the cached members of the groups. Also returns the updated cached members.
         """
+        if members is None:
+            members = self.members
+
+        try:
+            self._backing_group.members = members
+        except ObjectDoesNotExist:
+            StaticGroup.all_objects.create(
+                name=f"_{self.name}"[:CHARFIELD_MAX_LENGTH],
+                description=f"Backing static group for dynamic group {self.name}"[:CHARFIELD_MAX_LENGTH],
+                hidden=True,
+                content_type=self.content_type,
+                _dynamic_group=self,
+            )
+            self._backing_group.members = members
+        logger.debug("Refreshed cache for %s, now with %d members", self, self._backing_group.count)
 
-        cache.delete(self.members_cache_key)
+        # Since a change to a group or group of groups does not affect its children, we only need to go up the tree.
```

Comment:
The terminology here is admittedly confusing. A "parent" group is one that aggregates several "child" groups into a new group; i.e. group "A" (parent) is defined as "devices in group "B" (child 1) AND NOT group "C" (child 2)". So if we edit the definition of group C, that will also affect the membership calculation for group A; conversely, editing the definition of group A will have no direct effect on the membership of group C.

This block of code is relocated from the `dynamic_group_update_cached_members` signal handler as it seemed at the time like we should always do this when refreshing a given group, but maybe it would make sense to put it back there (i.e., only trigger parent refreshes when actually *editing* a child group, not when just refreshing it without changes)? I can do some testing and see how much of an impact it makes.